src/cargo/util/cache_lock.rs

```diff
@@ -0,0 +1,463 @@
+//! mode.
+//!
+//! An important rule is that `MutateExclusive` acquires the locks in the
+//! order "mutate" first and then the "download". That helps prevent
+//! deadlocks. It is not allowed for a cargo to first acquire a
+//! `DownloadExclusive` lock and then a `Shared` lock because that would open
+//! it up for deadlock.
+//!
+//! Another rule is that there should be only one [`CacheLocker`] per process
+//! to uphold the ordering rules. You could in theory have multiple if you
+//! could ensure that other threads would make progress and drop a lock, but
+//! cargo is not architected that way.
+//!
+//! It is safe to recursively acquire a lock as many times as you want.
+//!
+//! ## Interaction with older cargos
+//!
+//! Before version 1.74, cargo only acquired the `DownloadExclusive` lock when
+//! downloading and doing resolution. Newer cargos that acquire
+//! `MutateExclusive` should still correctly block when an old cargo is
+//! downloading (because it also acquires `DownloadExclusive`), but they do
+//! not properly coordinate when an old cargo is in the build phase (because
+//! it holds no locks). This isn't expected to be much of a problem because
+//! the intended use of mutating the cache is only to delete old contents
+//! which aren't currently being used. It is possible for there to be a
+//! conflict, particularly if the user manually deletes the entire cache, but
+//! it is not expected for this scenario to happen too often, and the only
+//! consequence is that one side or the other encounters an error and needs to
+//! retry.
+//!
+//! [^1]: A minor caveat is that downloads will delete an existing `src`
+//!   directory if it was extracted via an old cargo. See
+//!   [`crate::sources::registry::RegistrySource::unpack_package`]. This
+//!   should probably be fixed, but is unlikely to be a problem if the user is
+//!   only using versions of cargo with the same deletion logic.
+
+use super::FileLock;
+use crate::CargoResult;
+use crate::Config;
+use anyhow::Context;
+use std::cell::RefCell;
+use std::io;
+
+/// The style of lock to acquire.
+#[derive(Copy, Clone, Debug)]
+pub enum CacheLockMode {
+    /// A `Shared` lock allows multiple cargos to read from the source files.
+    ///
+    /// If another cargo has a `MutateExclusive` lock, then an attempt to get
+    /// a `Shared` will block.
+    ///
+    /// If another cargo has a `DownloadExclusive` lock, then the both can
+    /// operate concurrently under the assumption that downloading does not
+    /// modify existing source files.
+    Shared,
+    /// A `DownloadExclusive` lock ensures that only one cargo is downloading
+    /// new packages.
+    ///
+    /// If another cargo has a `MutateExclusive` lock, then an attempt to get
+    /// a `DownloadExclusive` lock will block.
+    ///
+    /// If another cargo has a `Shared` lock, then both can operate
+    /// concurrently.
+    DownloadExclusive,
```

Comment:
Seems like this should be the common / default choice.  Maybe should put it first.